packages/multicast_dns/lib/src/resource_record.dart

```diff
@@ -0,0 +1,182 @@
+// Copyright 2018 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:meta/meta.dart';
+
+/// Enumeration of support resource record types.
+class RRType {
+  /// An IPv4 Address record (1).
+  static const int a = 1;
+
+  /// An IPv6 Address record (28).
+  static const int aaaa = 28;
+
+  /// An IP Address reverse map record (12).
+  static const int ptr = 12;
+
+  /// An available service record (33).
+  static const int srv = 33;
+
+  /// A text record (16).
+  static const int txt = 16;
+
+  // TODO(dnfield): Can we support ANY in some meaningful way?
+  // /// A query for all records of all types known to the name server.
+  // static const int any = 255;
+
+  /// Asserts that a given int is a valid RRType.
+  static void debugAssertValid(int rrType) {
+    assert(rrType == a ||
+        rrType == aaaa ||
+        rrType == ptr ||
+        rrType == srv ||
+        rrType == txt);
+  }
+}
+
+/// Base implementation of DNS resource records (RRs).
+abstract class ResourceRecord {
+  /// Creates a new ResourceRecord.
+  ResourceRecord(this.rrValue, this.name, this.validUntil);
+
+  /// The FQDN for this record.
+  final String name;
+
+  /// The epoch time at which point this record is valid for in the cache.
+  final int validUntil;
+
+  /// The raw resource record value.  See [RRType] for supported values.
+  final int rrValue;
+
+  String get _additionalInfo;
+
+  @override
+  String toString() =>
+      '$runtimeType{$name, validUntil: ${DateTime.fromMillisecondsSinceEpoch(validUntil)}, $_additionalInfo}';
+
+  @override
+  bool operator ==(Object other) {
+    return other is ResourceRecord && _equals(other);
+  }
+
+  @protected
+  bool _equals(ResourceRecord other) {
+    return other.name == name &&
+        other.validUntil == validUntil &&
+        other.rrValue == rrValue;
+  }
+
+  @override
+  int get hashCode {
+    // TODO(dnfield): Probably should go with a real hashing function here
+    // when https://github.com/dart-lang/sdk/issues/11617 is figured out.
+    return toString().hashCode;
```

Comment:
we should create a real hashCode here, even if it's just xor'ing the hashcodes of the components